### PR TITLE
Reorganize crates

### LINKED ISSUE #1528: Reorganize crates in the repository
## Why
We have created many crates lately and we feel that we need to re-organize the folder structure to make it clearer.

## What
Define a clear separation between **"released"** crates and **"internal"** crates, by moving all utility/test crates in a specific folder (e.g `mithril-internal`).

## How

- [x] Move utility crates in a new folder (`internal`?) of the repository (`mithril-persistence`, `mithril-doc`, `mithril-doc-derive`, `mithril-build-script`)
- [x] Migrate `mithril-common` build script in `mithril-build-script`
- [x] Fix the crates table generation in the release notes to support "released" crates only
- [ ] Publish `mithril-build-script` in `crates.io`
- [x] The [API reference ](https://mithril.network/doc/manual/developer-docs/references) page on the documentation needs to be updated with new crates, and mention an access to `crates.io` and `npmjs.com` when appropriate.

### PR DESCRIPTION
**Pull request by Alenar**
## Content

This PR reorganize our crates by adding a new `internal` subfolder where the following crates have been moved:
- `mithril-build-script`
- `mithril-doc`
- `mithril-doc-derive`
- `mithril-persistence`
Also a makefile and a Readme were missing for some of those crates, this PR fill this gap.

Secondly this PR move the build script logic of `mithril-common` to `mithril-build-script` and enhance them to remove all the faulty conversions that were written in the generated code of the open api version management.

Thirdly it fix the crates table generation in the release notes to support "released" crates only (we switch from an excluding approach to an including one).

Lastly it update the [API reference page ](https://mithril.network/doc/manual/developer-docs/references) of the documentation website to reference the newest crates.

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [x] Update README file (if relevant)
  - [x] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## Comments
The build problems of the nix pipelines are also solved in this PR.

## Issue(s)
Closes #1528

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -74,6 +72,16 @@ This repository consists of the following parts:
...
+* [**Internal**](./internal): shared tools and api used by **Mithril** crates.
+
+  * [**Mithril build script**](./internal/mithril-build-script): A toolbox for Mithril crates that need a build scripts phase.
+
+  * [**Mithril doc**](./internal/mithril-doc): an api that generate markdown documentation of a crate command lines arguments.
+
+  * [**Mithril doc derive**](./internal/mithril-doc-derive): macro implementation for **Mithril Doc**.
+
+  * [**Mithril persistence**](./internal/mithril-persistence): the **persistence** library that is used by the **Mithril network** nodes.
+
```
Comment (jpraynaud) [inline README.md:84]:
```suggestion
* [**Internal**](./internal): the shared tools and API used by **Mithril** crates.

  * [**Mithril build script**](./internal/mithril-build-script): a toolbox for Mithril crates using a build scripts phase.

  * [**Mithril doc**](./internal/mithril-doc): an API that generates markdown documentation for a crate command lines arguments.

  * [**Mithril doc derive**](./internal/mithril-doc-derive): a macro implementation used by **Mithril doc**.

  * [**Mithril persistence**](./internal/mithril-persistence): the **persistence** library that is used by the **Mithril network** nodes.

```

File: internal/mithril-build-script/src/open_api.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    open_api_spec_files
+}
+
+fn read_version_from_open_api_spec_file(spec_file_path: PathBuf) -> OpenAPIVersionRaw {
+    let yaml_spec = fs::read_to_string(spec_file_path).unwrap();
+    let open_api: serde_yaml::Value = serde_yaml::from_str(&yaml_spec).unwrap();
+    open_api["info"]["version"].as_str().unwrap().to_owned()
+}
+
+/// Generate the `get_open_api_versions_mapping` function based on the openapi files
```
Comment (jpraynaud) [inline internal/mithril-build-script/src/open_api.rs:35]:
```suggestion
/// Generate the `get_open_api_versions_mapping` function based on the Open API files
```

File: internal/mithril-build-script/src/open_api.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+/// in the given folders.
+pub fn generate_open_api_versions_mapping(paths: &[&Path]) -> String {
+    let open_api_spec_files = list_all_open_api_spec_files(paths);
+    // Use a BTreeMap to guarantee the order
```
Comment (jpraynaud) [inline internal/mithril-build-script/src/open_api.rs:39]: Maybe this is more explanatory as the concern is to generate always the same code?
```suggestion
    // Use a BTreeMap to guarantee the deterministic code generation below
```

File: internal/mithril-doc/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Mithril-doc
+
+An api that generate markdown documentation of a crate command lines arguments.
```
Comment (jpraynaud) [inline internal/mithril-doc/README.md:3]:
```suggestion
An API that generates markdown documentation of a crate command lines arguments.
```

**Review by jpraynaud — APPROVED** [review]
LGTM :+1: 

I left some comments for typos fix.

**Review by dlachaumepalo — APPROVED** [review]

**Review by jpraynaud — APPROVED** [review]
LGTM :+1:

**Pull request merged**